### PR TITLE
Interactive music: Vertical mixing

### PR DESCRIPTION
**Pull request by ViMaSter**
## Summary
This PR introduces the ability to [vertically mix music](https://en.wikipedia.org/wiki/Adaptive_music#:~:text=Music.%5B5%5D-,Vertical%20re%2Dorchestration,-is%20the%20technique) via the `&PLAY_SONG_VARIANT` action line.  
It also addresses a volume issue with dialogue chirps when speeding up text.

### Vertical mixing
The game now supports two ways of playing music. [Documentation is updated accordingly.](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/8eab282d5567bf759627534881e72155172cde97/docs/methods/Audio.md#play_song_variant)
- **static music** refers to the style of music present in case 1
  - the name of a `.wav` file (and optional fading duration) is specified after `&PLAY_SONG`
    - **if no song is playing**, the song starts (and optionally is faded in based on the fading duration)
    - **if any song is already playing** (even if it is an identical song), that song stops or is faded out based on the fading duration, and the new song is started or faded in based on the fading duration
- **dynamic music** refers to vertically mixed music that consists of **a base track** and **a variant** playing at the same time; while a base track is playing the game can fade between variants:
  - the name of a `DynamicSongAsset`, variant thereof, and optional fading duration is specified after `&PLAY_SONG_VARIANT`
    - **if no song is playing**, the specified base track and variant start at the same time (and optionally are faded in based on the fading duration)
    - **if a static or different dynamic base track is playing**, that song stops or is faded out based on the fading duration, and the specified base track and variant start at the same time (and optionally are faded in based on the fading duration)
    - **NEW: if the same base track is already playing, but a different variant is specified**, the base track continues playing, the new variant is set to start from the same point of the currently playing variant and the currently playing variant is cross-faded into the new variant based on the specified fading duration

### Text speedup volume issue

This PR also addresses an issue where the dialogue chirp sound is louder when skipping dialogue. (Thanks @Zewatara!)  
Background: The chirp is played after a specific set of characters has been displayed. As more characters are displayed in a shorter amount of time when skipping dialogue, the sound is played multiple times at the same time, resulting in bigger overall volume.  
It is now possible to specify a relative volume the chirp will be set to, when text speedup is enabled: 
image

### Vertical mixing
1. Open and run the `/Assets/Demo/Scenes/DynamicAudio.unity` scene
2. Play through the short demo script
3. Notice parts of `You Burgie Burgie` fading between different arrangements

### Text speedup volume issue
1. Start the game from the main menu
2. Notice the volume of dialogue chirps decreasing, when speeding up text

## Additional information

- `[ ]` Changes UI
- `[x]` Introduces new feature
  - Vertical mixing via `&PLAY_SONG_VARIANT`
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:

### CODE REVIEW

File: unity-ggjj/Assets/Resources/Audio/Music/Static.meta [hunk 1]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (ViMaSter) [inline unity-ggjj/Assets/Resources/Audio/Music/Static.meta:1]: Misusing this file to add a bit of technical context: With the addition of dynamic music, I have moved existing tracks into `Music/Static` and newly-added vertically mixed tracks into `Music/Dynamic` subdirectories respectively. That way, the auto-generated documentation can pick them both up separately.

File: unity-ggjj/Assets/Scripts/Audio/DynamicMusicData.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Linq;
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "New Dynamic Music Data", menuName = "Dynamic Music/Dynamic Music Data")]
+public class DynamicMusicData : ScriptableObject
```
Comment (ViMaSter) [inline unity-ggjj/Assets/Scripts/Audio/DynamicMusicData.cs:6]: Each "dynamic song" gets a subfolder inside the `Music/Dynamic` directory. Inside that directory, we have
 - a `.asset` of type `DynamicMusicData`, which references...
   - the `.wav`-file representing the base track (`primary`), for example:
     - `Music/Dynamic/YouBurgieBurgie/YouBurgieBurgie-BaseA.wav` 
   - **and**
   - the `.wav`-files representing available variants (`secondaryVariants`), for example:
     - `Music/Dynamic/YouBurgieBurgie/YouBurgieBurgie-BaseB.wav`
     - `Music/Dynamic/YouBurgieBurgie/YouBurgieBurgie-Dan.wav`
     - `Music/Dynamic/YouBurgieBurgie/YouBurgieBurgie-Burgie.wav`
     - `Music/Dynamic/YouBurgieBurgie/YouBurgieBurgie-EvilBurgie.wav`

---

I've introduced this ScriptableObject to not limit us to specific naming conventions; technically, we could instead just parse all directories inside `Music/Dynamic` and look for variants at runtime instead. It just felt a bit odd to me.
Open for suggestions or alternatives!

File: unity-ggjj/Assets/Scripts/AudioController/AudioController.cs [hunk 3]
```diff
@@ -3,61 +3,61 @@
...
+    public void SetDialogueChirpVolume(float volume)
     {
-        if (IsCurrentlyPlayingMusic())
-        {
-            yield return FadeOutSongCoroutine(transitionTime / 2f);
-        }
-
-        SetCurrentTrack(song);
-        yield return _musicVolumeManager.Fade(_musicVolumeManager.MaximumVolume, transitionTime / 2f);
+        _dialogueChirpVolumeManager.AudioSource.volume = volume;
```
Comment (ViMaSter) [inline unity-ggjj/Assets/Scripts/AudioController/AudioController.cs:60]: I'm actually unsure about this implementation for two reasons: 
1. Before this PR, `VolumeManager` were always tied to a UI element that allowed the user to change the volume for anything played via that instance.
2. [For the initial implementation](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/35efefb6de16e8c909eed444c3a04a74d783910b/unity-ggjj/Assets/Scripts/AudioController/AudioController.cs#L38) @Zewatara and I decided to just use the `MaximumVolume`-property of the `VolumeManager`. Out of consistency I changed this to directly modify the `AudioSource` instead in a later commit.

Not sure what's the best way here, curious to hear some unbiased thoughts about it...

**Review by ViMaSter — COMMENTED** [review]
Added some comments to elaborate on and ask for feedback about specific parts of the implementation.

**Review by notexplosive — APPROVED** [review]

**Pull request merged**